In [40]:


import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [41]:
df_raw = pd.read_csv('fifa21 raw data v2.csv', low_memory=False)

df = df_raw.copy()

# Clean column names
df.columns = df.columns.str.replace('!', '', regex=False).str.strip()

In [42]:
df.shape


(18979, 77)

In [43]:
# Print all raw column names
print(df_raw.columns.tolist())

['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age', '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End', 'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle', 'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats', 'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY', 'Hits']


In [44]:
# Find any column names that contain the '↓' symbol
print([c for c in df_raw.columns if '↓' in c])

['↓OVA']


In [45]:
# Compare raw vs cleaned column names (first 5 only)
print("Before:", df_raw.columns.tolist()[:5])
print("After: ", df.columns.tolist()[:5])

Before: ['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl']
After:  ['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl']


In [46]:
# Show only the columns whose names changed after cleaning
changed = [(old, new) for old, new in zip(df_raw.columns, df.columns) if old != new]
print(changed)

[]


In [47]:
# 1. Size of the data (this also covers the shape check from Step 1)
print(df.shape)

# 2. First 5 rows (use a wide view so columns aren't hidden)
pd.set_option('display.max_columns', None)
df.head()

(18979, 77)


,ID,Name,LongName,photoUrl,playerUrl,Nationality,Age,↓OVA,POT,Club,Contract,Positions,Height,Weight,Preferred Foot,BOV,Best Position,Joined,Loan Date End,Value,Wage,Release Clause,Attacking,Crossing,Finishing,Heading Accuracy,Short Passing,Volleys,Skill,Dribbling,Curve,FK Accuracy,Long Passing,Ball Control,Movement,Acceleration,Sprint Speed,Agility,Reactions,Balance,Power,Shot Power,Jumping,Stamina,Strength,Long Shots,Mentality,Aggression,Interceptions,Positioning,Vision,Penalties,Composure,Defending,Marking,Standing Tackle,Sliding Tackle,Goalkeeping,GK Diving,GK Handling,GK Kicking,GK Positioning,GK Reflexes,Total Stats,Base Stats,W/F,SM,A/W,D/W,IR,PAC,SHO,PAS,DRI,DEF,PHY,Hits
0,158023,L. Messi,Lionel Messi,https://cdn.sofifa.com/players/158/023/21_60.png,http://sofifa.com/player/158023/lionel-messi/2...,Argentina,33,93,93,\n\n\n\nFC Barcelona,2004 ~ 2021,"RW, ST, CF",170cm,72kg,Left,93,RW,"Jul 1, 2004",NaN,€103.5M,€560K,€138.4M,429,85,95,70,91,88,470,96,93,94,91,96,451,91,80,91,94,95,389,86,68,72,69,94,347,44,40,93,95,75,96,91,32,35,24,54,6,11,15,14,8,2231,466,4 ★,4★,Medium,Low,5 ★,85,92,91,95,38,65,771
1,20801,Cristiano Ronaldo,C. Ronaldo dos Santos Aveiro,https://cdn.sofifa.com/players/020/801/21_60.png,http://sofifa.com/player/20801/c-ronaldo-dos-s...,Portugal,35,92,92,\n\n\n\nJuventus,2018 ~ 2022,"ST, LW",187cm,83kg,Right,92,ST,"Jul 10, 2018",NaN,€63M,€220K,€75.9M,437,84,95,90,82,86,414,88,81,76,77,92,431,87,91,87,95,71,444,94,95,84,78,93,353,63,29,95,82,84,95,84,28,32,24,58,7,11,15,14,11,2221,464,4 ★,5★,High,Low,5 ★,89,93,81,89,35,77,562
2,200389,J. Oblak,Jan Oblak,https://cdn.sofifa.com/players/200/389/21_60.png,http://sofifa.com/player/200389/jan-oblak/210006/,Slovenia,27,91,93,\n\n\n\nAtlético Madrid,2014 ~ 2023,GK,188cm,87kg,Right,91,GK,"Jul 16, 2014",NaN,€120M,€125K,€159.4M,95,13,11,15,43,13,109,12,13,14,40,30,307,43,60,67,88,49,268,59,78,41,78,12,140,34,19,11,65,11,68,57,27,12,18,437,87,92,78,90,90,1413,489,3 ★,1★,Medium,Medium,3 ★,87,92,78,90,52,90,150
3,192985,K. De Bruyne,Kevin De Bruyne,https://cdn.sofifa.com/players/192/985/21_60.png,http://sofifa.com/player/192985/kevin-de-bruyn...,Belgium,29,91,91,\n\n\n\nManchester City,2015 ~ 2023,"CAM, CM",181cm,70kg,Right,91,CAM,"Aug 30, 2015",NaN,€129M,€370K,€161M,407,94,82,55,94,82,441,88,85,83,93,92,398,77,76,78,91,76,408,91,63,89,74,91,408,76,66,88,94,84,91,186,68,65,53,56,15,13,5,10,13,2304,485,5 ★,4★,High,High,4 ★,76,86,93,88,64,78,207
4,190871,Neymar Jr,Neymar da Silva Santos Jr.,https://cdn.sofifa.com/players/190/871/21_60.png,http://sofifa.com/player/190871/neymar-da-silv...,Brazil,28,91,91,\n\n\n\nParis Saint-Germain,2017 ~ 2022,"LW, CAM",175cm,68kg,Right,91,LW,"Aug 3, 2017",NaN,€132M,€270K,€166.5M,408,85,87,62,87,87,448,95,88,89,81,95,453,94,89,96,91,83,357,80,62,81,50,84,356,51,36,87,90,92,93,94,35,30,29,59,9,9,15,15,11,2175,451,5 ★,5★,High,Medium,5 ★,91,85,86,94,36,59,595


In [48]:
# 3. Random sample of 10 rows (shows more variety than head)
df.sample(10)

,ID,Name,LongName,photoUrl,playerUrl,Nationality,Age,↓OVA,POT,Club,Contract,Positions,Height,Weight,Preferred Foot,BOV,Best Position,Joined,Loan Date End,Value,Wage,Release Clause,Attacking,Crossing,Finishing,Heading Accuracy,Short Passing,Volleys,Skill,Dribbling,Curve,FK Accuracy,Long Passing,Ball Control,Movement,Acceleration,Sprint Speed,Agility,Reactions,Balance,Power,Shot Power,Jumping,Stamina,Strength,Long Shots,Mentality,Aggression,Interceptions,Positioning,Vision,Penalties,Composure,Defending,Marking,Standing Tackle,Sliding Tackle,Goalkeeping,GK Diving,GK Handling,GK Kicking,GK Positioning,GK Reflexes,Total Stats,Base Stats,W/F,SM,A/W,D/W,IR,PAC,SHO,PAS,DRI,DEF,PHY,Hits
11233,148067,A. Prepeliță,Andrei Prepeliță,https://cdn.sofifa.com/players/148/067/21_60.png,http://sofifa.com/player/148067/andrei-prepeli...,Romania,34,64,64,\n\n\n\nFC Argeș,2019 ~ 2021,"CM, CDM",178cm,70kg,Right,66,CDM,"Sep 5, 2019",NaN,€275K,€2K,€297K,308,64,61,62,67,54,313,64,57,60,64,68,310,56,58,59,68,69,322,69,53,63,70,67,314,69,66,60,56,63,62,192,63,66,63,59,14,10,15,10,10,1818,380,3 ★,3★,Medium,Medium,1 ★,57,64,63,65,64,67,8
7217,213717,P. Forsell,Petteri Forsell,https://cdn.sofifa.com/players/213/717/21_60.png,http://sofifa.com/player/213717/petteri-forsel...,Finland,29,68,68,\n\n\n\nStal Mielec,2020 ~ 2021,"CAM, ST",168cm,72kg,Right,68,CAM,"Aug 10, 2020",NaN,€1.2M,€2K,€1.2M,305,66,70,42,68,59,356,66,75,75,69,71,329,56,54,60,73,86,346,77,63,56,74,76,264,46,23,65,71,59,63,151,46,54,51,63,11,12,15,11,14,1814,370,5 ★,3★,Medium,Low,1 ★,55,71,69,68,44,63,5
6776,259188,V. Jemelka,Václav Jemelka,https://cdn.sofifa.com/players/259/188/21_60.png,http://sofifa.com/player/259188/vaclav-jemelka...,Czech Republic,25,68,73,\n\n\n\nOud-Heverlee Leuven,2016 ~ 2022,CB,186cm,75kg,Left,70,CB,"Jul 1, 2016",NaN,€1.6M,€6K,€1.7M,252,51,33,71,66,31,220,41,37,20,68,54,313,60,65,59,70,59,262,41,63,66,70,22,258,73,71,25,51,38,68,203,67,70,66,50,7,9,9,11,14,1558,339,2 ★,2★,Medium,Medium,1 ★,63,32,57,49,69,69,3
3016,242429,L. Cacace,Liberato Cacace,https://cdn.sofifa.com/players/242/429/21_60.png,http://sofifa.com/player/242429/liberato-cacac...,New Zealand,19,73,84,\n\n\n\nSint-Truidense VV,2020 ~ 2023,"LB, LWB",184cm,72kg,Left,75,LM,"Aug 28, 2020",NaN,€6.5M,€6K,€9.9M,308,68,63,56,70,51,343,74,67,62,68,72,386,86,82,74,68,76,367,67,73,85,79,63,343,69,68,73,72,61,74,207,70,73,64,54,12,11,10,9,12,2008,437,4 ★,3★,High,Medium,1 ★,84,64,69,73,69,78,228
14630,251081,C. Gliga,Ciprian Gliga,https://cdn.sofifa.com/players/251/081/21_60.png,http://sofifa.com/player/251081/ciprian-gliga/...,Romania,23,61,67,\n\n\n\nAcademica Clinceni,2017 ~ 2024,"RM, RW",180cm,75kg,Right,62,RM,"Jul 1, 2017",NaN,€525K,€1K,€543K,246,60,56,30,58,42,269,58,38,61,51,61,342,79,76,73,52,62,257,57,47,64,44,45,227,28,44,55,55,45,47,77,20,29,28,48,9,9,12,9,9,1466,321,3 ★,3★,Medium,Medium,1 ★,77,53,56,60,29,46,1
16959,243353,B. Mumba,Bali Mumba,https://cdn.sofifa.com/players/243/353/21_60.png,http://sofifa.com/player/243353/bali-mumba/210...,England,18,57,77,\n\n\n\nNorwich City,2020 ~ 2024,"RB, CM",163cm,57kg,Right,57,RB,"Jul 27, 2020",NaN,€425K,€500,€580K,222,42,41,50,58,31,241,52,42,36,56,55,361,73,72,80,53,83,268,52,63,61,49,43,252,61,47,42,54,48,57,165,55,56,54,56,14,9,8,11,14,1565,333,3 ★,2★,High,Medium,1 ★,72,44,52,57,53,55,NaN
12574,242073,J. Blázquez,Joaquín Blázquez,https://cdn.sofifa.com/players/242/073/21_60.png,http://sofifa.com/player/242073/joaquin-blazqu...,Argentina,19,63,80,\n\n\n\nClub Atlético Talleres,2017 ~ 2022,GK,193cm,83kg,Right,63,GK,"Jun 1, 2017",NaN,€1.1M,€1K,€1.4M,60,10,6,11,26,7,79,9,11,10,33,16,181,21,25,33,57,45,188,47,65,22,47,7,78,19,12,7,26,14,33,35,10,12,13,312,63,59,63,63,64,933,335,3 ★,1★,Medium,Medium,1 ★,63,59,63,64,23,63,14
13184,213553,M. Ferrier,Morgan Ferrier,https://cdn.sofifa.com/players/213/553/21_60.png,http://sofifa.com/player/213553/morgan-ferrier...,England,25,62,65,\n\n\n\nTranmere Rovers,2019 ~ 2021,ST,183cm,75kg,Right,64,ST,"Aug 1, 2019",NaN,€525K,€2K,€684

In [49]:
# 4. Column names, data types, and non-null counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   ID                18979 non-null  int64
 1   Name              18979 non-null  str  
 2   LongName          18979 non-null  str  
 3   photoUrl          18979 non-null  str  
 4   playerUrl         18979 non-null  str  
 5   Nationality       18979 non-null  str  
 6   Age               18979 non-null  int64
 7   ↓OVA              18979 non-null  int64
 8   POT               18979 non-null  int64
 9   Club              18979 non-null  str  
 10  Contract          18979 non-null  str  
 11  Positions         18979 non-null  str  
 12  Height            18979 non-null  str  
 13  Weight            18979 non-null  str  
 14  Preferred Foot    18979 non-null  str  
 15  BOV               18979 non-null  int64
 16  Best Position     18979 non-null  str  
 17  Joined            18979 non-null  str  
 1

In [50]:
# 1. Count missing values per column (show only columns that have any)
nulls = df.isnull().sum()
print(nulls[nulls > 0])

Loan Date End    17966
Hits              2595
dtype: int64


In [51]:
# 2. Count fully duplicated rows
print(df.duplicated().sum())

0


In [52]:
# 3. Summary statistics for the numeric columns
df.describe()

,ID,Age,↓OVA,POT,BOV,Attacking,Crossing,Finishing,Heading Accuracy,Short Passing,Volleys,Skill,Dribbling,Curve,FK Accuracy,Long Passing,Ball Control,Movement,Acceleration,Sprint Speed,Agility,Reactions,Balance,Power,Shot Power,Jumping,Stamina,Strength,Long Shots,Mentality,Aggression,Interceptions,Positioning,Vision,Penalties,Composure,Defending,Marking,Standing Tackle,Sliding Tackle,Goalkeeping,GK Diving,GK Handling,GK Kicking,GK Positioning,GK Reflexes,Total Stats,Base Stats,PAC,SHO,PAS,DRI,DEF,PHY
count,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.00000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000,18979.000000
mean,226403.384794,25.194109,65.718636,71.136414,66.751726,248.938142,49.688392,45.842405,51.942726,58.768112,42.696507,256.479214,55.587491,47.246957,42.391011,52.725381,58.528373,317.718689,64.357553,64.412930,63.366721,61.655619,63.925865,296.605195,57.801676,64.577375,62.653986,64.786922,46.785236,254.199747,55.591285,46.394489,50.298962,53.857105,48.057906,58.07366,139.909268,46.559197,47.705622,45.644449,81.436693,16.416302,16.207124,16.076453,16.217187,16.519627,1595.286949,355.702197,67.453975,53.457031,57.681016,62.875020,49.866221,64.368934
std,27141.054157,4.710520,6.968999,6.114635,6.747193,74.299428,18.131153,19.567081,17.294409,14.519106,17.646937,78.650601,18.761314,18.207790,17.227947,15.178151,16.565892,55.879046,14.889981,14.638743,14.596277,9.072114,14.062285,50.732390,13.308747,11.844870,15.804223,12.488672,19.300534,64.595613,17.140414,20.698078,19.428701,13.708574,15.655999,12.07911,61.219472,20.139324,21.367690,20.922087,84.665579,17.554194,16.816305,16.491103,17.002239,17.854079,269.874789,40.761117,10.677859,13.827425,10.081857,9.927415,16.443213,9.601883
min,41.000000,16.000000,47.000000,47.000000,48.000000,42.000000,6.000000,3.000000,5.000000,7.000000,3.000000,40.000000,5.000000,4.000000,5.000000,5.000000,5.000000,122.000000,13.000000,12.000000,14.000000,24.000000,12.000000,122.000000,18.000000,15.000000,12.000000,16.000000,4.000000,50.000000,9.000000,3.000000,2.000000,9.000000,6.000000,12.00000,20.000000,3.000000,5.000000,4.000000,10.000000,2.000000,2.000000,2.000000,2.000000,2.000000,747.000000,232.000000,25.000000,16.000000,25.000000,25.000000,12.000000,28.000000
25%,210135.000000,21.000000,61.000000,67.000000,62.000000,222.000000,38.000000,30.000000,44.000000,54.000000,30.000000,222.000000,49.000000,35.000000,31.000000,43.000000,54.000000,289.000000,57.000000,57.000000,55.000000,56.000000,56.000000,264.000000,48.000000,58.000000,55.000000,57.000000,32.000000,227.500000,44.000000,25.000000,40.000000,45.000000,39.000000,50.00000,83.000000,29.000000,27.000000,25.000000,48.000000,8.000000,8.000000,8.000000,8.000000,8.000000,1452.000000,327.000000,61.000000,44.000000,51.000000,57.000000,35.000000,58.000000
50%,232418.000000,25.000000,66.000000,71.000000,67.000000,263.000000,54.000000,49.000000,55.000000,62.000000,44.000000,269.000000,61.000000,49.000000,41.000000,56.000000,63.000000,327.000000,67.000000,67.000000,66.000000,62.000000,66.000000,302.000000,59.000000,65.000000,66.000000,66.000000,51.000000,263.000000,58.000000,53.000000,55.000000,55.000000,49.000000,59.00000,159.000000,52.000000,55.000000,52.000000,53.000000,11.000000,11.000000,11.000000,11.000000,11.000000,1627.000000,356.000000,68.000000,56.000000,58.000000,64.000000,53.000000,65.000000
75%,246922.500000,29.000000,70.000000,75.000000,71.000000,297.

In [53]:
# 4. How many heights are in cm? (the rest are feet/inches)
print(df['Height'].str.endswith('cm').sum())

# 5. First 20 unique Hits values (look for K and NaN)
print(df['Hits'].unique()[:20])

18939
<ArrowStringArray>
[ '771',  '562',  '150',  '207',  '595',  '248',  '246',  '120', '1.6K',
  '130',  '321',  '189',  '175',   '96',  '118',  '216',  '212',  '154',
  '205',  '202']
Length: 20, dtype: str


In [54]:
# Missing values (only columns that have any)
nulls = df.isnull().sum()
print(nulls[nulls > 0])

# Duplicate rows
print("Duplicates:", df.duplicated().sum())

# Rows affected by each text problem
print("Height in feet/inches:", (~df['Height'].str.endswith('cm')).sum())
print("Weight in lbs:", df['Weight'].str.endswith('lbs').sum())
print("Hits with K:", df['Hits'].str.endswith('K', na=False).sum())
print("Value == 0:", (df['Value'] == '€0').sum())
print("Wage == 0:", (df['Wage'] == '€0').sum())

Loan Date End    17966
Hits              2595
dtype: int64
Duplicates: 0
Height in feet/inches: 40
Weight in lbs: 40
Hits with K: 28
Value == 0: 248
Wage == 0: 237


| Column | What is wrong | Rows affected |
|---|---|---|
| Club | Line breaks before the name | (we'll count this in step 2.1) |
| Value, Wage, Release Clause | € sign and M/K stored as text | 18,979 |
| Value / Wage | Zero values (€0) | 248 / 237 |
| Height | cm and feet/inches mixed | 40 in feet/inches |
| Weight | kg and lbs mixed | 40 in lbs |
| W/F, SM, IR | ★ symbol on the rating | 18,979 |
| Hits | Missing cells, K used for thousands | 2,595 missing, 28 with K |
| Joined | Date stored as text | 18,979 |
| Loan Date End | Mostly empty (normal) | 17,966 empty |
| Duplicates | Repeated rows | 0 |

In [55]:
# Show the raw club names exactly as stored (repr-style list shows the \n characters)
print(df['Club'].head(3).tolist())

# Count the Club cells that contain a line break
print("Clubs with line breaks:", df['Club'].str.contains('\n').sum())

['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus', '\n\n\n\nAtlético Madrid']
Clubs with line breaks: 18742


In [56]:
# Pick every text (object) column
text_cols = df.select_dtypes(include='object').columns

# For each one: turn line breaks into spaces, then trim spaces from both ends
for c in text_cols:
    df[c] = df[c].str.replace('\n', ' ', regex=False).str.strip()

C:\Users\Abubakar Sadiq\AppData\Local\Temp\ipykernel_1296\2838821759.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_cols = df.select_dtypes(include='object').columns


In [57]:
print(df['Club'].head(3).tolist())
print("Clubs with line breaks:", df['Club'].str.contains('\n').sum())
df[['Name', 'Club']].head()

['FC Barcelona', 'Juventus', 'Atlético Madrid']
Clubs with line breaks: 0


,Name,Club
0,L. Messi,FC Barcelona
1,Cristiano Ronaldo,Juventus
2,J. Oblak,Atlético Madrid
3,K. De Bruyne,Manchester City
4,Neymar Jr,Paris Saint-Germain


In [58]:
cols = ['Value', 'Wage', 'Release Clause']
print(df[cols].head())
print(df[cols].dtypes)

     Value   Wage Release Clause
0  €103.5M  €560K        €138.4M
1     €63M  €220K         €75.9M
2    €120M  €125K        €159.4M
3    €129M  €370K          €161M
4    €132M  €270K        €166.5M
Value             str
Wage              str
Release Clause    str
dtype: object


In [59]:
def convert_amount(x):
    """Convert text like '€67.5M', '€220K' or '€0' into a plain number."""
    if pd.isna(x):                      # keep missing cells as NaN
        return np.nan
    x = str(x).replace('€', '').strip() # remove the euro sign
    if x.endswith('M'):
        return round(float(x[:-1]) * 1_000_000)   # x[:-1] drops the last letter
    elif x.endswith('K'):
        return round(float(x[:-1]) * 1_000)
    else:
        return float(x)                 # plain number like '0'

In [60]:
for c in cols:
    df[c] = df[c].apply(convert_amount)

print(df[cols].head())
print(df[cols].dtypes)

         Value      Wage  Release Clause
0  103500000.0  560000.0     138400000.0
1   63000000.0  220000.0      75900000.0
2  120000000.0  125000.0     159400000.0
3  129000000.0  370000.0     161000000.0
4  132000000.0  270000.0     166500000.0
Value             float64
Wage              float64
Release Clause    float64
dtype: object


In [61]:
print("Value == 0:", (df['Value'] == 0).sum())
print("Wage == 0:", (df['Wage'] == 0).sum())

Value == 0: 248
Wage == 0: 237


In [62]:
# Look at the 40 feet/inches rows before changing anything
mask_cm = df['Height'].str.endswith('cm')
print(df.loc[~mask_cm, 'Height'].head())
print(df['Height'].dtype)

793     6'2"
847     6'3"
848     6'5"
860    5'11"
861     6'4"
Name: Height, dtype: str
str


In [63]:
# 1. Rows already in cm: pull out the digits before 'cm'
cm = df['Height'].str.extract(r'(\d+)cm')[0].astype(float)

# 2. Rows in feet/inches: pull out feet and inches as two separate numbers
ft_in = df['Height'].str.extract(r"(\d+)'(\d+)")
feet = ft_in[0].astype(float)
inches = ft_in[1].astype(float)

# 3. Convert feet/inches to cm using the formula from the assignment
ft_cm = feet * 30.48 + inches * 2.54

# 4. Use the cm value where it exists, otherwise the converted one; round to 1 decimal
df['Height'] = cm.fillna(ft_cm).round(1)

In [64]:
print(df['Height'].head())
print(df['Height'].dtype)
print("Missing:", df['Height'].isnull().sum())
print("Min:", df['Height'].min(), "Max:", df['Height'].max())

0    170.0
1    187.0
2    188.0
3    181.0
4    175.0
Name: Height, dtype: float64
float64
Missing: 0
Min: 155.0 Max: 206.0


In [65]:
# Look at the lbs rows before changing anything
is_lbs = df['Weight'].str.endswith('lbs')
print(df.loc[is_lbs, 'Weight'].head())
print(df['Weight'].dtype)

793    183lbs
847    179lbs
848    183lbs
860    172lbs
861    196lbs
Name: Weight, dtype: str
str


In [66]:
# True for rows stored in pounds
is_lbs = df['Weight'].str.endswith('lbs')

# Pull the digits out of each format (the other format gives NaN)
kg  = df['Weight'].str.extract(r'(\d+)kg')[0].astype(float)
lbs = df['Weight'].str.extract(r'(\d+)lbs')[0].astype(float)

# If the row is in lbs, convert with lbs * 0.453592; otherwise keep the kg value
df['Weight'] = np.where(is_lbs, lbs * 0.453592, kg).round(1)

In [67]:
print(df['Weight'].head())
print(df['Weight'].dtype)
print("Missing:", df['Weight'].isnull().sum())
print("Min:", df['Weight'].min(), "Max:", df['Weight'].max())

0    72.0
1    83.0
2    87.0
3    70.0
4    68.0
Name: Weight, dtype: float64
float64
Missing: 0
Min: 50.0 Max: 110.0


In [68]:
star_cols = ['W/F', 'SM', 'IR']
print(df[star_cols].head())
print(df[star_cols].dtypes)
print(df['W/F'].unique())


   W/F  SM   IR
0  4 ★  4★  5 ★
1  4 ★  5★  5 ★
2  3 ★  1★  3 ★
3  5 ★  4★  4 ★
4  5 ★  5★  5 ★
W/F    str
SM     str
IR     str
dtype: object
<ArrowStringArray>
['4 ★', '3 ★', '5 ★', '2 ★', '1 ★']
Length: 5, dtype: str


In [69]:
for c in star_cols:
    # Grab the first digit (r'(\d)') and drop the star symbol
    df[c] = df[c].str.extract(r'(\d)')[0]
    # Turn the extracted text digit into a real number
    df[c] = pd.to_numeric(df[c])

In [70]:
print(df[star_cols].head())
print(df[star_cols].dtypes)
print("Missing:", df[star_cols].isnull().sum().sum())
print(df['W/F'].unique())

   W/F  SM  IR
0    4   4   5
1    4   5   5
2    3   1   3
3    5   4   4
4    5   5   5
W/F    int64
SM     int64
IR     int64
dtype: object
Missing: 0
[4 3 5 2 1]


In [71]:
print(df['Hits'].head(10).tolist())
print("Missing:", df['Hits'].isnull().sum())
print("With K:", df['Hits'].str.endswith('K', na=False).sum())
print(df['Hits'].dtype)

['771', '562', '150', '207', '595', '248', '246', '120', '1.6K', '130']
Missing: 2595
With K: 28
str


In [72]:
# convert_amount already handles K, plain numbers and NaN
df['Hits'] = df['Hits'].apply(convert_amount)

In [73]:
median_hits = df['Hits'].median()
print("Median:", median_hits)

df['Hits'] = df['Hits'].fillna(median_hits)

Median: 5.0


In [74]:
print(df['Hits'].head(10).tolist())
print(df['Hits'].dtype)
print("Missing:", df['Hits'].isnull().sum())
print(df.loc[df['Hits'] >= 1000, 'Hits'].head())   # the former K rows

[771.0, 562.0, 150.0, 207.0, 595.0, 248.0, 246.0, 120.0, 1600.0, 130.0]
float64
Missing: 0
8      1600.0
29     1100.0
34     1500.0
68     3200.0
105    1700.0
Name: Hits, dtype: float64


In [75]:
print(df['Joined'].head().tolist())
print(df['Joined'].dtype)

['Jul 1, 2004', 'Jul 10, 2018', 'Jul 16, 2014', 'Aug 30, 2015', 'Aug 3, 2017']
str


In [76]:
# %b = short month name (Jul), %d = day, %Y = 4-digit year
df['Joined'] = pd.to_datetime(df['Joined'], format='%b %d, %Y')

In [77]:
# Reference date: 1 January 2021 (the assignment's choice)
ref_date = pd.Timestamp('2021-01-01')

# Difference in days, divided by 365.25 to get years (365.25 accounts for leap years)
df['Years_at_Club'] = ((ref_date - df['Joined']).dt.days / 365.25).round(1)

In [78]:
print(df[['Joined', 'Years_at_Club']].head())
print(df['Joined'].dtype)
print("Missing:", df['Joined'].isnull().sum())

# Players with more than 10 years at their club
long_serving = df[df['Years_at_Club'] > 10][['Name', 'Club', 'Joined', 'Years_at_Club']]
print("Players with more than 10 years:", len(long_serving))
long_serving.sort_values('Years_at_Club', ascending=False)

      Joined  Years_at_Club
0 2004-07-01           16.5
1 2018-07-10            2.5
2 2014-07-16            6.5
3 2015-08-30            5.3
4 2017-08-03            3.4
datetime64[us]
Missing: 0
Players with more than 10 years: 206


,Name,Club,Joined,Years_at_Club
10264,H. Sogahata,Kashima Antlers,1998-01-01,23.0
501,I. Akinfeev,PFC CSKA Moscow,2002-01-01,19.0
4770,Kim Kwang Suk,Pohang Steelers,2002-01-01,19.0
9948,D. Lewington,Milton Keynes Dons,2002-08-01,18.4
6480,K. Nakamura,Kawasaki Frontale,2003-01-01,18.0
...,...,...,...,...
1750,I. Pillud,Racing Club,2010-08-01,10.4
3370,Z. Erdal,Sivasspor,2010-09-01,10.3
16064,M. Haeder,SC Verl,2010-11-01,10.2
4402,Zhang Linpeng,Guangzhou Evergrande Taobao FC,2010-11-22,10.1


In [79]:
#Outliers
# Q1 = 25% mark, Q3 = 75% mark of the Wage values
q1, q3 = df['Wage'].quantile([0.25, 0.75])
iqr = q3 - q1                     # the spread of the middle 50% of players

lower_fence = q1 - 1.5 * iqr
upper_fence = q3 + 1.5 * iqr

print("Q1:", q1, "Q3:", q3, "IQR:", iqr)
print("Lower fence:", lower_fence)
print("Upper fence:", upper_fence)

Q1: 1000.0 Q3: 8000.0 IQR: 7000.0
Lower fence: -9500.0
Upper fence: 18500.0


In [80]:
# True if the wage is outside either fence, otherwise False
df['Wage_Outlier'] = (df['Wage'] < lower_fence) | (df['Wage'] > upper_fence)

print(df['Wage_Outlier'].value_counts())
print("Rows still in df:", len(df))

Wage_Outlier
False    16623
True      2356
Name: count, dtype: int64
Rows still in df: 18979


In [81]:
df.nlargest(10, 'Wage')[['Name', 'Club', 'Wage']]

,Name,Club,Wage
0,L. Messi,FC Barcelona,560000.0
3,K. De Bruyne,Manchester City,370000.0
15,K. Benzema,Real Madrid,350000.0
26,E. Hazard,Real Madrid,350000.0
12,Casemiro,Real Madrid,310000.0
27,T. Kroos,Real Madrid,310000.0
16,Sergio Ramos,Real Madrid,300000.0
17,S. Agüero,Manchester City,300000.0
28,A. Griezmann,FC Barcelona,290000.0
4,Neymar Jr,Paris Saint-Germain,270000.0


In [82]:
# Typical player vs the biggest earner
print("Median wage:", df['Wage'].median())
print("Max wage:", df['Wage'].max())
print("Upper fence:", upper_fence)
print("Players flagged as outliers:", df['Wage_Outlier'].sum())

Median wage: 3000.0
Max wage: 560000.0
Upper fence: 18500.0
Players flagged as outliers: 2356


In [83]:
# Duplicates
# Count duplicates now that the data is cleaned
print("Duplicates before:", df.duplicated().sum())
rows_before = len(df)

# Drop repeated rows (keeps the first copy of each)
df = df.drop_duplicates()

rows_after = len(df)
print("Rows before:", rows_before)
print("Rows after: ", rows_after)
print("Rows removed:", rows_before - rows_after)

Duplicates before: 0
Rows before: 18979
Rows after:  18979
Rows removed: 0


In [84]:
print(df['Loan Date End'].isnull().sum())

17966


In [85]:
print(df['Contract'].sample(10, random_state=1).tolist())

['2020 ~ 2024', '2020 ~ 2022', '2020 ~ 2021', '2019 ~ 2020', '2020 ~ 2024', '2020 ~ 2024', '2019 ~ 2024', '2014 ~ 2022', '2018 ~ 2022', '2019 ~ 2023']


In [86]:
treated = ['Value', 'Wage', 'Height', 'Weight', 'Hits']

# 1. No missing values in the cleaned columns
assert df[treated].isnull().sum().sum() == 0

# 2. Every cleaned column is numeric
for c in treated:
    assert pd.api.types.is_numeric_dtype(df[c]), c

# 3. Heights and weights are in realistic ranges
assert df['Height'].between(150, 220).all()
assert df['Weight'].between(50, 120).all()

# 4. No rows were lost compared to the raw data
assert len(df) == len(df_raw)

print("All Stage 3 checks passed")

All Stage 3 checks passed


In [87]:
# Star columns are plain numbers between 1 and 5
for c in ['W/F', 'SM', 'IR']:
    assert df[c].between(1, 5).all(), c

# Joined is a real date, and the new columns exist
assert pd.api.types.is_datetime64_any_dtype(df['Joined'])
assert 'Years_at_Club' in df.columns and 'Wage_Outlier' in df.columns
print("Extra checks passed")

Extra checks passed


In [88]:
# Rows that do NOT follow the 'year ~ year' pattern
other = df.loc[~df['Contract'].str.contains('~'), 'Contract']
print("Rows with another format:", len(other))
print(other.unique()[:10])

Rows with another format: 1250
<ArrowStringArray>
['Jun 30, 2021 On Loan',                 'Free', 'Dec 31, 2020 On Loan',
 'Jan 30, 2021 On Loan', 'Jun 30, 2022 On Loan', 'May 31, 2021 On Loan',
  'Jul 5, 2021 On Loan', 'Dec 31, 2021 On Loan',  'Jul 1, 2021 On Loan',
  'Jan 1, 2021 On Loan']
Length: 10, dtype: str


In [89]:
# Save the cleaned dataframe (index=False avoids an extra index column)
df.to_csv('fifa21_cleaned_data.csv', index=False)

# Read it back to check that the types survived
pd.read_csv('fifa21_cleaned_data.csv').info()

<class 'pandas.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 79 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                18979 non-null  int64  
 1   Name              18979 non-null  str    
 2   LongName          18979 non-null  str    
 3   photoUrl          18979 non-null  str    
 4   playerUrl         18979 non-null  str    
 5   Nationality       18979 non-null  str    
 6   Age               18979 non-null  int64  
 7   ↓OVA              18979 non-null  int64  
 8   POT               18979 non-null  int64  
 9   Club              18979 non-null  str    
 10  Contract          18979 non-null  str    
 11  Positions         18979 non-null  str    
 12  Height            18979 non-null  float64
 13  Weight            18979 non-null  float64
 14  Preferred Foot    18979 non-null  str    
 15  BOV               18979 non-null  int64  
 16  Best Position     18979 non-null  str    
 17  Join